feat(scopes): Add project:distribution permission scope

runningcode opened the pull request:
## Summary

Adds a new permission scope `project:distribution` for Sentry Build Distribution functionality. This scope allows checking for updates of a given mobile app distribution from the mobile app.

IMPORTANT: We want this scope to be available for integration tokens but not personal tokens or org tokens. This token will be embedded in a mobile app that is distributed and therefore we want to minimize the possibility of using any other token type with more scope than needed in case it is leaked.

## Changes

- Added `project:distribution` to `ApiScopes.project` tuple in `src/sentry/models/apiscopes.py`
- Added `project:distribution` to `ScopesDict` TypedDict
- Added `project:distribution` to `SENTRY_SCOPES` in `src/sentry/conf/server.py`
- Added `project:distribution` to `SENTRY_SCOPE_HIERARCHY_MAPPING` in `src/sentry/conf/server.py`
- Generated database migrations to update BitField scopes for API token models

runningcode committed:
feat(scopes): Add project:distribution permission scope

Adds a new permission scope for Sentry Distribution functionality.

Changes:
- Added project:distribution to ApiScopes.project tuple
- Added project:distribution to ScopesDict TypedDict
- Added project:distribution to SENTRY_SCOPES
- Added project:distribution to SENTRY_SCOPE_HIERARCHY_MAPPING
- Added project:distribution to SENTRY_SCOPE_SETS with description
- Updated tests to verify the new scope works correctly

runningcode commented on src/sentry/conf/server.py:
```diff
@@ -1728,6 +1728,7 @@ def custom_parameter_sort(parameter: dict) -> tuple[str, int]:
     "project:write",
     "project:admin",
     "project:releases",
+    "project:distribution",
```
So apparently this is the set of all possible scopes so we need to add it here.

runningcode committed:
refactor(scopes): Remove project:distribution from SENTRY_SCOPE_SETS

runningcode commented on src/sentry/models/apiscopes.py:
```diff
@@ -21,7 +21,13 @@ def add_scope_hierarchy(curr_scopes: Sequence[str]) -> list[str]:
+        ("project:admin"),
+        ("project:releases"),
+        ("project:distribution"),
+    )
```
This seems wrong. Looking at the migrations, it looks like it is the last item in the bitfield.

runningcode committed:
feat(scopes): Add project:distribution to SENTRY_SCOPES and hierarchy mapping

Registers the project:distribution scope in SENTRY_SCOPES and
SENTRY_SCOPE_HIERARCHY_MAPPING to make it available for permission
checks throughout the system.

runningcode committed:
refactor(scopes): Remove project:distribution scope migrations and type definition

Remove the database migrations that added project:distribution to scope
bitfields and remove the scope from the ScopesDict TypedDict definition.

runningcode commented on src/sentry/migrations/0999_add_distribution_token.py:
```diff
@@ -0,0 +1,149 @@
+    operations = [
+        migrations.AlterField(
+            model_name="apiauthorization",
+            name="scopes",
```
This was just auto-generated. Let me delete and rerun the migrations.

runningcode replied:
This is still there after rerunning the migrations. Do you suggest I manually delete it?

runningcode committed:
test(scopes): Mark project:distribution as non-default owner scope

The project:distribution scope should not be included in the default
set of scopes granted to organization owners, similar to org:ci and
other specialized scopes.

michelletran-sentry reviewed (approved):
I'm approving to unblock. I _think_ that the bitfield is actually unused (so probably don't need to add the scope to the `ApiScopes` sequence), but I can follow this up myself.

runningcode committed:
fix(migration): Fix migration dependency to point to master

The migration was depending on 0999 which doesn't exist on master yet,
causing the migration drift workflow to fail. Updated to depend on 0997,
which is the latest migration on master.

runningcode committed:
fix(migration): Update dependency to 0999 after rebase

After rebasing on master, migration 0999 now exists, so migration 1000
needs to depend on 0999 instead of 0997 to avoid conflicting leaf nodes
in the migration graph.

runningcode committed:
fix(migrations): Support migration numbers >= 1000 in squash script

The squash script was only detecting migrations starting with '0', which
broke when we reached migration 1000. Update to match any file starting
with a digit.

chromy reviewed (approved):
lgtm
